### PR TITLE
⚠️ Bump to controller-runtime v0.17.0 & k8s.io v0.29

### PR DESCRIPTION
**Pull request by sbueringer**
Signed-off-by: Stefan Büringer buringerst@vmware.com

**What this PR does / why we need it**:
I've split the PR intentionally in separate commits, to make the follow-up tasks after the bump clear and to make it easier to review.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Part of #9578

### CODE REVIEW

**Comment by sbueringer** [comment]
/assign @fabriziopandini @chrischdi 

Ready for review now

**Comment by fabriziopandini** [comment]
Great work!
/lgtm

What about adding a comment to the changes in API diff into the migration v16-->1.7 doc

**Comment by kashifest** [comment]
lgtm
Thanks @sbueringer , would this be backported to release-1.6 branch or ?

**Comment by sbueringer** [comment]
> lgtm Thanks @sbueringer , would this be backported to release-1.6 branch or ?

No. We consider minor bumps of controller-runtime and Kubernetes libraries a breaking change, so we won't backport them.

**Comment by kashifest** [comment]
> No. We consider minor bumps of controller-runtime and Kubernetes libraries a breaking change, so we won't backport them.

Makes sense, thanks

File: internal/controllers/machinedeployment/mdutil/util_test.go [hunk 1]
```diff
@@ -30,12 +30,16 @@ import (
...
 	"k8s.io/apiserver/pkg/storage/names"
-	"k8s.io/klog/v2/klogr"
 	"k8s.io/utils/ptr"
+	ctrl "sigs.k8s.io/controller-runtime"
 
 	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
 )
 
+var (
+	ctx = ctrl.SetupSignalHandler()
```
Comment (chrischdi) [inline internal/controllers/machinedeployment/mdutil/util_test.go:40]: Note: we won't have a logger in this context, right? (Propably similar for other places, but I only see tests affected (if even))

File: internal/controllers/machinedeployment/machinedeployment_sync_test.go [hunk 2]
```diff
@@ -577,7 +574,7 @@ func TestComputeDesiredMachineSet(t *testing.T) {
 		expectedMS := skeletonMSBasedOnMD.DeepCopy()
 
 		g := NewWithT(t)
-		actualMS, err := (&Reconciler{}).computeDesiredMachineSet(deployment, nil, nil, log)
+		actualMS, err := (&Reconciler{}).computeDesiredMachineSet(ctx, deployment, nil, nil)
```
Comment (chrischdi) [inline internal/controllers/machinedeployment/machinedeployment_sync_test.go:577]: Also this context may not have a logger right?

File: internal/test/envtest/environment.go [hunk 3]
```diff
@@ -72,11 +71,10 @@ import (
 )
 
 func init() {
-	klog.InitFlags(nil)
-	logger := klogr.New()
+	logger := klog.Background()
```
Comment (chrischdi) [inline internal/test/envtest/environment.go:74]: Wouldn't the replacement be `textlogger.NewLogger` ?

**Review by chrischdi — COMMENTED** [review]
Just some questions / nits around klogr.New replacements. lgtm otherwise

File: internal/controllers/machinedeployment/mdutil/util_test.go (hunk 1 above)
Comment (sbueringer) [reply internal/controllers/machinedeployment/mdutil/util_test.go:40]: There is no logger, but ctrl.LoggerFrom has a fallback for this case. Although the fallback isn't logging anything if there is no global logger set 

See: https://github.com/kubernetes-sigs/controller-runtime/blob/main/pkg/log/log.go#L91-L99
(NullLogSink)

File: internal/test/envtest/environment.go (hunk 3 above)
Comment (sbueringer) [reply internal/test/envtest/environment.go:74]: That is what the godoc is saying, yup. I though it's better to use the same in envtest that we also use in main.go

Comment (sbueringer) [reply internal/test/envtest/environment.go:74]: It falls back to some logger so it is not the same as in main.go, but I thought it doesn't make much sense / wouldn't actually help much if I add all the log flag stuff we have in main.go for this init func of a test util (as we would never set the flags)

xref: https://github.com/kubernetes-sigs/cluster-api/blob/8d4a510d3b21798829c93c753c8880e511bfe8ed/main.go#L239-L242

Basically ValidateAndApply reads the flags and then configures some logger. This logger is then used through klog.Background()

File: internal/controllers/machinedeployment/machinedeployment_sync_test.go (hunk 2 above)
Comment (sbueringer) [reply internal/controllers/machinedeployment/machinedeployment_sync_test.go:577]: Not sure if it's necessary to ensure a global logger is set so LoggerFrom is falling back to some real logger

File: internal/test/envtest/environment.go (hunk 3 above)
Comment (chrischdi) [reply internal/test/envtest/environment.go:74]: Fine for me 👍

File: internal/controllers/machinedeployment/mdutil/util_test.go (hunk 1 above)
Comment (chrischdi) [reply internal/controllers/machinedeployment/mdutil/util_test.go:40]: Fine for me 👍

File: internal/controllers/machinedeployment/machinedeployment_sync_test.go (hunk 2 above)
Comment (chrischdi) [reply internal/controllers/machinedeployment/machinedeployment_sync_test.go:577]: Fine for me 👍

**Review by chrischdi — COMMENTED** [review]
/lgtm

**Comment by fabriziopandini** [comment]
@sbueringer it seems that all the comments are addressed
are we good to go now?

**Comment by sbueringer** [comment]
@fabriziopandini Yes we are!

**Comment by fabriziopandini** [comment]
yay!
/lgtm
/approve

**Pull request merged**